lib/resourcemerge/core.go

```diff
@@ -254,16 +255,19 @@ func EnsureServicePorts(modified *bool, existing *[]corev1.ServicePort, required
 }
 
 func ensureServicePort(modified *bool, existing *corev1.ServicePort, required corev1.ServicePort) {
-	ensureServicePortDefaults(&required)
+	ensureServicePortDefaults(&required, existing)
 	if !equality.Semantic.DeepEqual(required, *existing) {
 		*modified = true
 		*existing = required
 	}
 }
 
-func ensureServicePortDefaults(servicePort *corev1.ServicePort) {
-	if servicePort.Protocol == "" {
-		servicePort.Protocol = corev1.ProtocolTCP
+func ensureServicePortDefaults(required, existing *corev1.ServicePort) {
+	if required.Protocol == "" {
+		required.Protocol = corev1.ProtocolTCP
+	}
+	if required.TargetPort == intstr.FromInt(0) {
+		required.TargetPort = intstr.FromInt(int(required.Port))
```

Comment:
Can we use:

```go
	if servicePort.TargetPort == intstr.FromInt(0) || servicePort.TargetPort == intstr.FromString("") {
		servicePort.TargetPort = intstr.FromInt(int(servicePort.Port))
```

to match [the upstream default logic][1]?  Actually vendoring the upstream default logic would be another option, but feels a bit heavy to me, if we plan on moving to server-side-apply soonish.  But copy/pasting the relevant upstream default code seems safer to me than writing something that's similar but not quite identical locally.

[1]: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/apis/core/v1/defaults.go#L122-L124

Reply:
Good idea, added this in c37b1c74.